File: tests/test_code/ruby/mass_assignment.rb
# Ruby on Rails Mass Assignment
class UsersController < ApplicationController
  def create
    # VULNERABLE: Direct params passed to create
    @user = User.create(params[:user]) # Line 5
  end

  def update
    @user = User.find(params[:id])
    # VULNERABLE: Direct params passed to update
    @user.update(params[:user]) # Line 11
  end
end

File: tests/test_code/python/mass_assignment.py
from django import forms
from .models import User

# VULNERABLE: Django Mass Assignment via Form
class UserForm(forms.ModelForm):
    class Meta:
        model = User
        fields = '__all__' # Line 7

# VULNERABLE: Direct Mass Assignment in View
def update_user(request, user_id):
    user = User.objects.get(id=user_id)
    # Taint flows from request.POST to save()
    user.save(request.POST) # Line 13
